kolibri/core/assets/src/exams/utils.js

```diff
@@ -84,106 +88,113 @@ function annotateQuestionsWithItem(questions) {
  */
-export function convertExamQuestionSourcesToV3(exam, extraArgs = {}) {
-  if (exam.data_model_version !== 3) {
-    const V2_sources = convertExamQuestionSources(exam, extraArgs);
-    return [convertV2toV3(V2_sources, exam)];
-  }
-
-  return exam.question_sources;
-}
 
-/**
- * @returns V2 formatted question_sources
- */
-export function convertExamQuestionSources(exam, extraArgs = {}) {
-  const { data_model_version } = exam;
-  if (data_model_version === 0) {
-    // TODO contentNodes are only needed for V0 -> V2 conversion, but a request to the
-    // ContentNode API is made regardless of the version being converted
-    if (extraArgs.contentNodes === undefined) {
-      throw new Error(
-        "Missing 'contentNodes' array, which is required when converting a V0 Exam model"
-      );
-    }
-    if (exam.seed === undefined) {
-      throw new Error("Missing 'seed' integer, which is required when converting a V0 Exam model");
-    }
-    const { contentNodes } = extraArgs;
-    const questionIds = {};
-    contentNodes.forEach(node => {
-      questionIds[node.id] = node.assessmentmetadata
-        ? node.assessmentmetadata.assessment_item_ids
-        : [];
+export async function convertExamQuestionSources(exam) {
+  if (exam.data_model_version === 0) {
+    const ids = uniq(exam.question_sources.map(item => item.exercise_id));
+    const exercises = await ContentNodeResource.fetchCollection({
+      getParams: {
+        ids,
+        no_available_filtering: true,
+      },
     });
-    return annotateQuestionsWithItem(
-      convertExamQuestionSourcesV0V2(exam.question_sources, exam.seed, questionIds)
+    const questionIds = exercises.reduce((nodeIds, node) => {
+      nodeIds[node.id] = node.assessmentmetadata ? node.assessmentmetadata.assessment_item_ids : [];
+      return nodeIds;
+    }, []);
+    exam.question_sources = convertExamQuestionSourcesV0V2(
+      exam.question_sources,
+      exam.seed,
+      questionIds
     );
+    // v1 -> v2 only updates the `counter_in_exercise` field if it's in camelCase
+    // so we can set the data_model_version to 2 here to skip that code
+    exam.data_model_version = 2;
   }
-  if (data_model_version === 1) {
-    return annotateQuestionsWithItem(convertExamQuestionSourcesV1V2(exam.question_sources));
+
+  if (exam.data_model_version === 1) {
+    exam.question_sources = convertExamQuestionSourcesV1V2(exam.question_sources);
+    exam.data_model_version = 2;
   }
 
-  // For backwards compatibility. If you are using V3, use the convertExamQuestionSourcesToV3 func
-  if (data_model_version === 3) {
-    return revertV3toV2(exam.question_sources);
+  if (exam.data_model_version === 2) {
+    exam.question_sources = convertExamQuestionSourcesV2toV3(exam);
+    exam.data_model_version = 3;
   }
 
-  return annotateQuestionsWithItem(exam.question_sources);
+  // TODO This avoids updating older code that used `item` to refer to the unique question id
```

Comment:
Looks like this needs to be updated/removed in line with this issue, which is binding us to the opposite course of action: https://github.com/learningequality/kolibri/issues/12127

Maybe we should just update the annotateQuestionsWithItem to operate on a question_sources in the v3 shape?